tests/build_stitched_index.cpp

```diff
@@ -0,0 +1,855 @@
+    std::ofstream universal_label_writer;
+    universal_label_writer.exceptions(std::ios::badbit | std::ios::failbit);
+    universal_label_writer.open(final_index_path_prefix +
+                                "_universal_label.txt");
+    universal_label_writer << universal_label << std::endl;
+    universal_label_writer.close();
+  }
+
+  // main index
+  _u64         index_num_frozen_points = 0, index_num_edges = 0;
+  _u32         index_max_observed_degree = 0, index_entry_point = 0;
+  const size_t METADATA = 2 * sizeof(_u64) + 2 * sizeof(_u32);
+  for (auto &point_neighbors : stitched_graph) {
+    index_max_observed_degree =
+        std::max(index_max_observed_degree, (_u32) point_neighbors.size());
+  }
+
+  std::ofstream stitched_graph_writer;
+  stitched_graph_writer.exceptions(std::ios::badbit | std::ios::failbit);
+  stitched_graph_writer.open(final_index_path_prefix);
+
+  stitched_graph_writer.write((char *) &final_index_size, sizeof(_u64));
+  stitched_graph_writer.write((char *) &index_max_observed_degree,
+                              sizeof(_u32));
+  stitched_graph_writer.write((char *) &index_entry_point, sizeof(_u32));
+  stitched_graph_writer.write((char *) &index_num_frozen_points, sizeof(_u64));
+
+  size_t bytes_written = METADATA;
+  for (_u32 node_point = 0; node_point < stitched_graph.size(); node_point++) {
+    _u32 current_node_num_neighbors = stitched_graph[node_point].size();
+    std::vector<_u32> current_node_neighbors = stitched_graph[node_point];
+    stitched_graph_writer.write((char *) &current_node_num_neighbors,
+                                sizeof(_u32));
+    bytes_written += sizeof(_u32);
+    for (const auto &current_node_neighbor : current_node_neighbors) {
+      stitched_graph_writer.write((char *) &current_node_neighbor,
+                                  sizeof(_u32));
+      bytes_written += sizeof(_u32);
+    }
+    index_num_edges += current_node_num_neighbors;
+  }
+
+  if (bytes_written != final_index_size) {
+    std::cerr << "Error: written bytes does not match allocated space"
+              << std::endl;
+    throw;
+  }
+
+  stitched_graph_writer.close();
+
+  std::chrono::duration<double> saving_index_time =
+      std::chrono::high_resolution_clock::now() - saving_index_timer;
+  std::cout << "Stitched graph written in " << saving_index_time.count()
+            << " seconds" << std::endl;
+  std::cout << "Stitched graph average degree: "
+            << ((float) index_num_edges) / ((float) (stitched_graph.size()))
+            << std::endl;
+  std::cout << "Stitched graph max degree: " << index_max_observed_degree
+            << std::endl
+            << std::endl;
+}
+
+/*
+ * Unions the per-label graph indices together via the following policy:
+ *  - any two nodes can only have at most one edge between them -
+ *
+ * Returns the "stitched" graph and its expected file size.
+ */
+template<typename T>
+stitch_indices_return_values stitch_label_indices(
```

Comment:
Could many of these helper functions, not just this one, be moved to common utils files so that they can be reused by future code? Maybe have a src/filter_utils.cpp. Placing them here means they will not get reused in the future.

Reply:
While I'm not convinced as to the reusability of this code, I think that the level of refactoring required would take much more time and should constitute a subsequent PR on top of this one.